apps/zbugs/server/notify.ts

```diff
@@ -202,20 +258,80 @@ function clip(s: string) {
 async function gatherRecipients(
   tx: ServerTransaction<Schema, postgres.TransactionSql>,
   issueID: string,
+  isAssigneeChange = false,
+  previousAssigneeID?: string | null | undefined,
 ): Promise<string[]> {
   const sql = tx.dbTransaction.wrappedTransaction;
 
-  const recipientRows = await sql`SELECT DISTINCT "user".email
-  FROM (
-      SELECT DISTINCT "creatorID" AS "userID" FROM "comment" WHERE "issueID" = ${issueID}
+  // Get all recipient candidates
+  const recipientRows = await sql`
+    WITH issue_info AS (
+      SELECT 
+        "creatorID",
+        "assigneeID",
+        "visibility"
+      FROM "issue"
+      WHERE id = ${issueID}
+    ),
+    recipient_candidates AS (
+      -- Issue creator
+      SELECT DISTINCT "user".id, "user".email, "user".role
+      FROM issue_info
+      JOIN "user" ON "user".id = issue_info."creatorID"
+      
+      UNION
+      
+      -- Users with emojis on the issue
+      SELECT DISTINCT "user".id, "user".email, "user".role
+      FROM "emoji"
+      JOIN "user" ON "user".id = "emoji"."creatorID"
+      WHERE "emoji"."subjectID" = ${issueID}
+      
+      UNION
+      
+      -- Users who have commented
+      SELECT DISTINCT "user".id, "user".email, "user".role
+      FROM "comment"
+      JOIN "user" ON "user".id = "comment"."creatorID"
+      WHERE "comment"."issueID" = ${issueID}
+      
       UNION
-      SELECT DISTINCT "creatorID" AS "userID" FROM "emoji" WHERE "subjectID" = ${issueID}
+      
+      -- Users with emojis on comments
+      SELECT DISTINCT "user".id, "user".email, "user".role
+      FROM "emoji"
+      JOIN "comment" ON "comment".id = "emoji"."subjectID"
+      JOIN "user" ON "user".id = "emoji"."creatorID"
+      WHERE "comment"."issueID" = ${issueID}
+      
       UNION
-      SELECT "creatorID" AS "userID" FROM "issue" WHERE "id" = ${issueID}
+      
+      -- Current assignee (if exists)
+      SELECT DISTINCT "user".id, "user".email, "user".role
+      FROM issue_info
+      JOIN "user" ON "user".id = issue_info."assigneeID"
+      WHERE issue_info."assigneeID" IS NOT NULL
+      
       UNION
-      SELECT "assigneeID" AS "userID" FROM "issue" WHERE "id" = ${issueID}
-  ) AS combined
-  JOIN "user" ON "user".id = combined."userID" AND "user".email IS NOT NULL;`;
+      
+      -- Previous assignee (if this is an assignee change and previous assignee exists)
+      SELECT DISTINCT "user".id, "user".email, "user".role
+      FROM "user"
+      WHERE ${isAssigneeChange} 
+        AND ${previousAssigneeID ? sql`"user".id = ${previousAssigneeID}` : sql`FALSE`}
+    )
+    SELECT DISTINCT email
+    FROM recipient_candidates
+    WHERE email IS NOT NULL
+    AND (
+      -- If issue is public, include all candidates
+      EXISTS (
+        SELECT 1 FROM issue_info WHERE visibility = 'public'
+      )
+      OR
+      -- If issue is not public, only include crew members
+      role = 'crew'
```

Comment:
I'd write a test for this query just to be sure we don't end up emailing people on internal issues.